File: app/src/main/java/com/team2052/frckrawler/views/metric/impl/SliderMetricWidget.java
package com.team2052.frckrawler.views.metric.impl;

import android.content.Context;
import android.support.v7.widget.AppCompatSeekBar;
import android.widget.TextView;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.jakewharton.rxbinding.widget.RxSeekBar;
import com.team2052.frckrawler.R;
import com.team2052.frckrawler.database.metric.MetricValue;
import com.team2052.frckrawler.tba.JSON;
import com.team2052.frckrawler.util.MetricHelper;
import com.team2052.frckrawler.views.metric.MetricWidget;

import rx.android.schedulers.AndroidSchedulers;

public class SliderMetricWidget extends MetricWidget {

    private final AppCompatSeekBar seekBar;
    int value;
    private int min;
    private int max;
    private TextView valueText;

    public SliderMetricWidget(Context context, MetricValue metricValue) {
        super(context, metricValue);
        inflater.inflate(R.layout.widget_metric_slider, this);
        seekBar = (AppCompatSeekBar) findViewById(R.id.sliderVal);
        setMetricValue(metricValue);
    }

    public SliderMetricWidget(Context context) {
        super(context);
        inflater.inflate(R.layout.widget_metric_slider, this);
        seekBar = (AppCompatSeekBar) findViewById(R.id.sliderVal);
        valueText = (TextView) findViewById(R.id.value);

        RxSeekBar.userChanges(seekBar).subscribeOn(AndroidSchedulers.mainThread()).subscribe(seekValue -> {
            valueText.setText(Integer.toString(seekValue));
            value = seekValue + min;
        });
    }

    @Override
    public void setMetricValue(MetricValue m) {
        min = 0;
        max = 1;
        ((TextView) findViewById(R.id.name)).setText(m.getMetric().getName());

        JsonObject range = JSON.getAsJsonObject(m.getMetric().getData());
        min = range.get("min").getAsInt();
        max = range.get("max").getAsInt();

        seekBar.setMax(max - min);

        ((TextView) findViewById(R.id.min)).setText(Integer.toString(min));
        ((TextView) findViewById(R.id.max)).setText(Integer.toString(max));

        if (m.getValue() != null && !m.getValue().getAsJsonObject().get("value").isJsonNull())
            value = m.getValue().getAsJsonObject().get("value").getAsInt();
        else
            value = min;

        if (value < min || value > max)
            value = min;
        seekBar.setProgress(value - min);
        valueText.setText(Integer.toString(value));
    }

    @Override
    public JsonElement getData() {
        return MetricHelper.buildNumberMetricValue(value);
    }
}

File: app/src/main/java/com/team2052/frckrawler/metric/view/impl/SliderMetricWidget.kt
package com.team2052.frckrawler.metric.view.impl

import android.content.Context
import android.support.v7.widget.AppCompatSeekBar
import android.view.View
import android.widget.TextView

import com.google.gson.JsonElement
import com.jakewharton.rxbinding.widget.RxSeekBar
import com.team2052.frckrawler.R
import com.team2052.frckrawler.data.tba.v3.JSON
import com.team2052.frckrawler.helpers.metric.MetricDataHelper
import com.team2052.frckrawler.metric.data.MetricValue
import com.team2052.frckrawler.metric.view.MetricWidget

import rx.android.schedulers.AndroidSchedulers

class SliderMetricWidget : MetricWidget {
    internal var value: Int = 0
    private var min: Int = 0
    private var max: Int = 0
    private var seekBar: AppCompatSeekBar? = null
    private var valueText: TextView? = null
    private var nameText: TextView? = null
    private var minText: TextView? = null
    private var maxText: TextView? = null

    @Suppress("unused")
    constructor(context: Context, metricValue: MetricValue) : super(context, metricValue) {
        setMetricValue(metricValue)
    }

    constructor(context: Context) : super(context) {}

    override fun setMetricValue(m: MetricValue) {
        min = 0
        max = 1
        nameText!!.text = m.metric.name

        val range = JSON.getAsJsonObject(m.metric.data)
        min = range.get("min").asInt
        max = range.get("max").asInt

        seekBar!!.max = max - min

        minText!!.text = Integer.toString(min)
        maxText!!.text = Integer.toString(max)

        if (m.value != null && !m.value.asJsonObject.get("value").isJsonNull)
            value = m.value.asJsonObject.get("value").asInt
        else
            value = min

        if (value < min || value > max)
            value = min
        seekBar!!.progress = value - min
        valueText!!.text = Integer.toString(value)
    }

    override fun initViews() {
        inflater.inflate(R.layout.widget_metric_slider, this)
        seekBar = findViewById<View>(R.id.sliderVal) as AppCompatSeekBar
        valueText = findViewById<View>(R.id.value) as TextView
        nameText = findViewById<View>(R.id.name) as TextView
        minText = findViewById<View>(R.id.min) as TextView
        maxText = findViewById<View>(R.id.max) as TextView

        RxSeekBar.userChanges(seekBar!!).subscribeOn(AndroidSchedulers.mainThread()).subscribe { seekValue ->
            value = seekValue!! + min
            valueText!!.text = Integer.toString(value)
        }
    }

    override val data: JsonElement get() = MetricDataHelper.buildNumberMetricValue(value)
}
